hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/state/ContractStorageReadableKVState.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright (C) 2024 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.mirror.web3.state;
+
+import com.hedera.hapi.node.state.contract.SlotKey;
+import com.hedera.hapi.node.state.contract.SlotValue;
+import com.hedera.mirror.common.domain.entity.EntityId;
+import com.hedera.mirror.web3.common.ContractCallContext;
+import com.hedera.mirror.web3.repository.ContractStateRepository;
+import com.hedera.pbj.runtime.io.buffer.Bytes;
+import com.swirlds.state.spi.ReadableKVStateBase;
+import java.util.Collections;
+import java.util.Iterator;
+import org.jetbrains.annotations.NotNull;
+
+public class ContractStorageReadableKVState extends ReadableKVStateBase<SlotKey, SlotValue> {
+
+    private static final String KEY = "STORAGE";
+    private final ContractStateRepository contractStateRepository;
+
+    protected ContractStorageReadableKVState(final ContractStateRepository contractStateRepository) {
+        super(KEY);
+        this.contractStateRepository = contractStateRepository;
+    }
+
+    @Override
+    protected SlotValue readFromDataSource(@NotNull SlotKey slotKey) {
+        if (!slotKey.hasContractID()) {
+            return null;
+        }
+
+        final var timestamp = ContractCallContext.get().getTimestamp();
+        final var contractID = slotKey.contractID();
+        final var entityId = EntityId.of(contractID.shardNum(), contractID.realmNum(), contractID.contractNum());
+        return timestamp
+                .map(t -> contractStateRepository.findStorageByBlockTimestamp(
+                        entityId.getId(), slotKey.key().toByteArray(), t))
+                .orElse(contractStateRepository.findStorage(
+                        entityId.getId(), slotKey.key().toByteArray()))
```

Comment:
nit: Can pull up some method calls to clean it up a bit.

```suggestion
        final var entityId = EntityId.of(contractID.shardNum(), contractID.realmNum(), contractID.contractNum()).getId();
        final var keyBytes = slotKey.key().toByteArray();

        return timestamp
                .map(t -> contractStateRepository.findStorageByBlockTimestamp(entityId, keyBytes, t))
                .orElse(contractStateRepository.findStorage(entityId, keyBytes))
```

Reply:
Comment addressed